magefile.go

```diff
@@ -147,260 +254,341 @@ func (Test) Clean() error {
+			"CGO_ENABLED": "0",
+		},
+		command.WithArgs{
+			"build", "-a",
+			"-o", filepath.Join(_projectRoot, "bin", "addon-metadata-operator"),
+			filepath.Join("cmd", "addon-metadata-operator", "main.go"),
+		},
+		command.WithConsoleOut(mg.Verbose()),
+		command.WithContext{Context: ctx},
 	)
+
+	if err := build.Run(); err != nil {
+		return fmt.Errorf("starting to build amo: %w", err)
+	}
+
+	if build.Success() {
+		return nil
+	}
+
+	return fmt.Errorf("building amo: %w", build.Error())
 }
 
+var gocmd = command.NewCommandAlias(mg.GoCmd())
+
 func (Build) CleanOperator() error {
 	return sh.Rm(filepath.Join(_projectRoot, "bin", "addon-metadata-operator"))
 }
 
-func (Build) OperatorImage() error {
-	runtime := runtime()
-	if runtime == "" {
+func (Build) OperatorImage(ctx context.Context) error {
+	runtime, ok := container.Runtime()
+	if !ok {
 		return errors.New("could not find container runtime")
 	}
 
-	return sh.Run(runtime, "build",
-		"-t", fmt.Sprintf("%s:%s", repository, _tag),
-		"-f", "Dockerfile.build", _projectRoot,
+	build := command.NewCommand(runtime,
+		command.WithArgs{
+			"build",
+			"-t", fmt.Sprintf("%s:%s", repository, _tag),
+			"-f", "Dockerfile.build", _projectRoot,
+		},
+		command.WithConsoleOut(mg.Verbose()),
+		command.WithContext{Context: ctx},
 	)
+
+	if err := build.Run(); err != nil {
+		return fmt.Errorf("starting to build amo image: %w", err)
+	}
+
+	if build.Success() {
+		return nil
+	}
+
+	return fmt.Errorf("building amo image: %w", build.Error())
 }
 
 type Generate mg.Namespace
 
 func (Generate) Manifests(ctx context.Context) error {
 	mg.CtxDeps(ctx, Deps.UpdateControllerGen)
 
-	return sh.Run(filepath.Join(_depBin, "controller-gen"),
-		"crd", "rbac:roleName=manager-role",
-		"webhook", `paths="./..."`,
-		"output:crd:artifacts:config=config/crd/bases",
+	generate := controllergen(
+		command.WithArgs{
+			"crd", "rbac:roleName=manager-role",
+			"webhook", `paths="./..."`,
+			"output:crd:artifacts:config=config/crd/bases",
+		},
+		command.WithConsoleOut(mg.Verbose()),
+		command.WithContext{Context: ctx},
 	)
+
+	if err := generate.Run(); err != nil {
+		return fmt.Errorf("starting to generate manifests: %w", err)
+	}
+
+	if generate.Success() {
+		return nil
+	}
+
+	return fmt.Errorf("generating manifests: %w", generate.Error())
 }
 
 func (Generate) Boilerplate(ctx context.Context) error {
 	mg.CtxDeps(ctx, Deps.UpdateControllerGen)
 
-	return sh.Run(filepath.Join(_depBin, "controller-gen"),
-		`object:headerFile="hack/boilerplate.go.txt"`,
-		`paths="./..."`,
+	generate := controllergen(
+		command.WithArgs{
+			`object:headerFile="hack/boilerplate.go.txt"`,
+			`paths="./..."`,
+		},
+		command.WithConsoleOut(mg.Verbose()),
+		command.WithContext{Context: ctx},
 	)
+
+	if err := generate.Run(); err != nil {
+		return fmt.Errorf("starting to generate boilerplate: %w", err)
+	}
+
+	if generate.Success() {
+		return nil
+	}
+
+	return fmt.Errorf("generating boilerplate: %w", generate.Error())
```

Comment:
nit: this seems like a repeatable pattern that might be abstracted. Just brainstorming but we might avoid the extra Success/Error check by modifying the Run method of the command interface: 

```go
// Run executes a "Command" instance and returns an error if
// either the command was not able to start.
func (c *Command) Run() error {
	if err := c.cmd.Run(); err != nil {
		if _, ok := err.(*exec.ExitError); !ok { //nolint:errorlint
			return fmt.Errorf("running command %q: %w", strings.Join(c.cmd.Args, " "), err)
		}
	}

	if !c.cmd.ProcessState.Success() {
            return fmt.Errof("command return error %q: %w", strings.Join(c.cmd.Args, " "), c.cmd.ProcessState.String())
        }

        return nil
        }
}
```

Reply:
(not going to paste this comment all over but pretty much all commands use that pattern)

Reply:
The issue is there still needs to be a way to distinguish between a failure to launch (e.g. a missing binary) or a process failure. The examples in this `magefile` are a little expanded because I'm wrapping the error.

So this:
```
if generate.Success() {
	return nil
}

return fmt.Errorf("generating boilerplate: %w", generate.Error())
```

Could just be:
```
return generate.Error()
```

This was actually one of the issues with the `magefile/sh` abstractions. It was difficult to react to failures since every error looks the same.

Reply:
That makes sense. Although you could convey the difference in the error message returned by `Run`

Reply:
I just feel as a developer/user of this CI library, I don't really care about the error type, as long as it's providing enough information for me to debug. I don't see the value in the extra error splitting complexity. But that's just opinion.

Reply:
I just feel as a developer/user of this CI library, I don't really care about the error type, as long as it's providing enough information for me to debug. I don't see the value in the extra error splitting complexity. But that's just opinion.